photo/models.py

```diff
@@ -122,9 +130,36 @@ def validate_unique(self, *args, **kwargs):
         qs = ContestSubmission.objects.filter(
             contest=self.contest, picture__user=self.picture.user
         )
+
         if qs.exists() and self._state.adding:
-            raise ValidationError("Each user can only submit one picture per contest")
+            raise ValidationError(UNIQUE_SUBMISSION_ERROR_MESSAGE)
+
+    def validate_vote(self):
+        user_vote = ContestSubmission.objects.filter(
+            contest=self.contest, votes=self.picture.user
+        )
+
+        if user_vote.exists() and self._state.adding:
+            raise ValidationError(REPEATED_VOTE_ERROR_MESSAGE)
+
+    def validate_submission_date(self):
+        if self.contest.upload_phase_end is not None and (
+            not (
+                self.contest.upload_phase_start
+                <= self.submission_date
+                <= self.contest.upload_phase_end
+            )
+        ):
+            raise ValidationError(OUTDATED_SUBMISSION_ERROR_MESSAGE)
 
     def save(self, *args, **kwargs):
         self.validate_unique()
+        self.validate_vote()
+        self.validate_submission_date()
         super(ContestSubmission, self).save(*args, **kwargs)
+
+    def add_vote(self, user):
+        if user not in self.votes.all():
```

Comment:
this check needs to be improved, all() is going to load everything into memory, doesn't make much sense.

Reply:
it will load every vote regarding that contest submission if I'm not mistaken